github-actions/unified-status-check/src/main.ts

```diff
@@ -0,0 +1,194 @@
+          return {
+            state: checkOrStatus.state,
+            name: checkOrStatus.context,
+          };
+        }
+        if (checkOrStatus.__typename === 'CheckRun') {
+          return {
+            state: checkOrStatus.conclusion
+              ? checkConclusionStateToStatusStateMap.get(checkOrStatus.conclusion)!
+              : 'PENDING',
+            name: checkOrStatus.name || 'Unknown Check Run',
+          };
+        }
+        throw Error();
+      })
+      // Do not check the unified status for its state as it can prevent itself from ever passing.
+      .filter(({name}) => name !== unifiedStatusCheckName);
+
+    const counts = statuses.reduce(
+      (counts, {state}) => {
+        if (isPassingState(state)) {
+          counts.passing += 1;
+        }
+        if (isPendingState(state)) {
+          counts.pending += 1;
+        }
+        if (isFailingState(state)) {
+          counts.failing += 1;
+        }
+        return counts;
+      },
+      {passing: 0, failing: 0, pending: 0},
+    );
+
+    if (counts.failing > 0) {
+      await github.repos.createCommitStatus({
+        ...context.repo,
+        sha: pullRequest.commits.nodes[0].commit.oid,
+        context: unifiedStatusCheckName,
+        state: 'failure',
+        description: `${counts.failing} expected status(es) failing`,
+      });
+      return;
+    }
+
+    if (counts.pending > 0) {
+      await github.repos.createCommitStatus({
+        ...context.repo,
+        state: 'pending',
+        sha: pullRequest.commits.nodes[0].commit.oid,
+        context: unifiedStatusCheckName,
+        description: 'Expected statuses are still pending',
+      });
+      return;
+    }
+
+    await github.repos.createCommitStatus({
+      ...context.repo,
+      sha: pullRequest.commits.nodes[0].commit.oid,
+      context: unifiedStatusCheckName,
+      state: 'success',
+    });
+  } finally {
+    await revokeActiveInstallationToken(github);
+  }
+}
+
+/** Whether a status state is a passing state. */
+function isPassingState(state: StatusState) {
+  return state === 'SUCCESS' || state === 'EXPECTED';
+}
+
+/** Whether a status state is a pending state. */
+function isPendingState(state: StatusState) {
+  return state === 'PENDING';
+}
+
+/** Whether a status state is a failing state. */
+function isFailingState(state: StatusState) {
+  return state === 'ERROR' || state === 'FAILURE';
+}
+
+/** Mapping of Github Check Conclusion states to Status states. */
+const checkConclusionStateToStatusStateMap = new Map<CheckConclusionState, StatusState>([
+  ['ACTION_REQUIRED', 'PENDING'],
+  ['CANCELLED', 'ERROR'],
+  ['FAILURE', 'FAILURE'],
+  ['NEUTRAL', 'EXPECTED'],
+  ['SKIPPED', 'EXPECTED'],
+  ['STALE', 'ERROR'],
+  ['STARTUP_FAILURE', 'FAILURE'],
+  ['SUCCESS', 'SUCCESS'],
+  ['TIMED_OUT', 'FAILURE'],
+]);
+
+/** GraphQL schema for requesting the status information for a given pull request. */
+const PR_SCHEMA = params(
```

Comment:
isn't all of this similar in ng-dev? can we just use that from there or dedupe?

Reply:
We could, but to do so we would have to use the `AuthenticatedGithubApi` class that is in `ng-dev`, which is more complicated because of the access to the configuration files.

Reply:
The client does not depend on configs AFAIK: https://github.com/angular/dev-infra/blob/main/ng-dev/utils/git/github.ts

Reply:
I was thinking of the `AuthenticatedGitClient` rather than the Github one.  Will see what I can do.

Reply:
From looking into it more, we can leverage the ng-dev stuff, but we don't really save ourselves that much code since we still have to describe what the shape of the PR response we want is.

Reply:
I thought we are already using a similar schema (almost identical, maybe with a bit more stuff). If that's not the case, then yes. I'm fine keeping it here. Maybe want to extract it into a separate file for more readability? it's odd having this at the end of a file too (not relying on hoisting for better readability)

Reply:
I will do the refactor into another file in a follow up because I want to expand a bit of what its doing and I would rather land this as is for now.

Reply:
no worries, sgtm